### ISSUE TITLE
[BUG]: Log error entry: "KeyError: 'regulation_mode'"

### ISSUE THREAD

**Issue by Xirt** [plugwise/python-plugwise#240]
### Describe the bug.

I get the following error in the log (both in the Sep (2022.9.4) and Nov (2022.11.1) version of HA):

```
Logger: homeassistant.components.select
Source: components/plugwise/select.py:113
Integration: Selecteer (documentation, issues)
First occurred: 10:00:35 (2 occurrences)
Last logged: 10:00:35

Error adding entities for domain select with platform plugwise
Error while setting up plugwise platform for select
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 428, in async_add_entities
    await asyncio.gather(*tasks)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 692, in _async_add_entity
    await entity.add_to_platform_finish()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 788, in add_to_platform_finish
    await self.async_added_to_hass()
  File "/usr/src/homeassistant/homeassistant/components/plugwise/entity.py", line 81, in async_added_to_hass
    self._handle_coordinator_update()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 348, in _handle_coordinator_update
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 545, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 583, in _async_write_ha_state
    state = self._stringify_state(available)
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 551, in _stringify_state
    if (state := self.state) is None:
  File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 97, in state
    if self.current_option is None or self.current_option not in self.options:
  File "/usr/src/homeassistant/homeassistant/components/plugwise/select.py", line 113, in current_option
    return self.device[self.entity_description.current_option_key]
KeyError: 'regulation_mode'
```

In addition, I have the feeling that the sync of information is not "near perfect" anymore. For example, if I set a new mode (e.g. no_frost, away, ...) from HA then the temperature is not always properly synchronized back to HA. In my example, I first set the thermostat in HA to no_frost (configured as 10C in Plugwise) and afterwards to away (configured as 17.5C in Plugwise). In the Plugwise app this both shows fine, but the thermostat eventually shows away / 10C. Restarting HA fixes the desync temporarily. Obviously there might be a relation hence I propose to first fix the error in the log :).

### Expected behavior

No errors and a good sync between Plugwise and HA.

### What version of Home Assistant Core has the issue?

2022.11.1

### What type of installation are you running?

Home Assistant Container

### How and which version of the integration do you have installed?

Home Assistant Core

### What kind of Plugwise device is having issues?

Smile: Adam (including Lisa, Tom, Floor)

### What firmware version is your Plugwise product at?

3.7.0

### Home Assistant Logging (optional)

2022-11-06 10:14:54.308 ERROR (MainThread) [homeassistant.components.select] Error adding entities for domain select with platform plugwise
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 428, in async_add_entities
await asyncio.gather(*tasks)
File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 692, in _async_add_entity
await entity.add_to_platform_finish()
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 788, in add_to_platform_finish
await self.async_added_to_hass()
File "/usr/src/homeassistant/homeassistant/components/plugwise/entity.py", line 81, in async_added_to_hass
self._handle_coordinator_update()
File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 348, in _handle_coordinator_update
self.async_write_ha_state()
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 545, in async_write_ha_state
self._async_write_ha_state()
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 583, in _async_write_ha_state
state = self._stringify_state(available)
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 551, in _stringify_state
if (state := self.state) is None:
File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 97, in state
if self.current_option is None or self.current_option not in self.options:
File "/usr/src/homeassistant/homeassistant/components/plugwise/select.py", line 113, in current_option
return self.device[self.entity_description.current_option_key]
KeyError: 'regulation_mode'
2022-11-06 10:14:54.312 ERROR (MainThread) [homeassistant.components.select] Error while setting up plugwise platform for select
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 289, in _async_setup_platform
await asyncio.gather(*pending)
File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 428, in async_add_entities
await asyncio.gather(*tasks)
File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 692, in _async_add_entity
await entity.add_to_platform_finish()
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 788, in add_to_platform_finish
await self.async_added_to_hass()
File "/usr/src/homeassistant/homeassistant/components/plugwise/entity.py", line 81, in async_added_to_hass
self._handle_coordinator_update()
File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 348, in _handle_coordinator_update
self.async_write_ha_state()
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 545, in async_write_ha_state
self._async_write_ha_state()
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 583, in _async_write_ha_state
state = self._stringify_state(available)
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 551, in _stringify_state
if (state := self.state) is None:
File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 97, in state
if self.current_option is None or self.current_option not in self.options:
File "/usr/src/homeassistant/homeassistant/components/plugwise/select.py", line 113, in current_option
return self.device[self.entity_description.current_option_key]
KeyError: 'regulation_mode'

### Diagnostics information

In case additional information is required, please let me know.

### Additional information

Running Plugwise with an Adam, 4x Smile and 1x Anna. They are connected to a Intergas Kombi Kompakt HRE 36/30A
CW5.

**Comment by bouwew** [comment]
Can you download the diagnostics and post these in a message?
Go to Settings --> Devices & Services --> Plugwise Adam --> 3 vertical points --> Download diagnostics

**Comment by bouwew** [comment]
On the sync of information, this is a known problem. HA expects an update from the Adam within 1.5 seconds, and the Adam doesn't always succeed to be updated in that time-frame. So sometimes/often in HA the change is toggles back to the previous value. But at the next update, 60 seconds later, the change is correctly shown.

This is not only the case for Plugwise, other integrations show similar behavior. When there is a final solution, it will be implemented in Plugwise, when it's not a generic solution.

**Comment by Xirt** [comment]
Find below the diagnostics:

```
{
  "home_assistant": {
    "installation_type": "Home Assistant Container",
    "version": "2022.11.1",
    "dev": false,
    "hassio": false,
    "virtualenv": false,
    "python_version": "3.10.7",
    "docker": true,
    "arch": "x86_64",
    "timezone": "Europe/Amsterdam",
    "os_name": "Linux",
    "os_version": "4.4.180+",
    "run_as_root": true
  },
  "custom_components": {},
  "integration_manifest": {
    "domain": "plugwise",
    "name": "Plugwise",
    "documentation": "https://www.home-assistant.io/integrations/plugwise",
    "requirements": [
      "plugwise==0.25.3"
    ],
    "codeowners": [
      "@CoMPaTech",
      "@bouwew",
      "@brefra",
      "@frenck"
    ],
    "zeroconf": [
      "_plugwise._tcp.local."
    ],
    "config_flow": true,
    "iot_class": "local_polling",
    "integration_type": "hub",
    "loggers": [
      "crcmod",
      "plugwise"
    ],
    "is_built_in": true
  },
  "data": {
    "gateway": {
      "smile_name": "Adam",
      "gateway_id": "removed",
      "heater_id": "removed",
      "cooling_present": false,
      "notifications": {}
    },
    "devices": {
      "92ddebd7c54a4afebf45572c5bf4509e": {
        "dev_class": "thermostat",
        "location": "a98e17e70adf42ae88801bfdc1cd2a3e",
        "model": "ThermoTouch",
        "name": "Thermostaat - Woonkamer",
        "vendor": "Plugwise",
        "thermostat": {
          "setpoint": 17.5,
          "lower_bound": 0.0,
          "upper_bound": 30.0,
          "resolution": 0.01
        },
        "available": true,
        "preset_modes": [
          "home",
          "vacation",
          "no_frost",
          "asleep",
          "away"
        ],
        "active_preset": "asleep",
        "available_schedules": [
          "Thermostat schedule"
        ],
        "selected_schedule": "None",
        "last_used": "Thermostat schedule",
        "control_state": "off",
        "mode": "heat",
        "sensors": {
          "temperature": 21.4,
          "setpoint": 17.5
        }
      },
      "cc990ab96cb4434c80cf4bc4b8fba7e3": {
        "dev_class": "thermo_sensor",
        "firmware": "2020-11-04T01:00:00+01:00",
        "hardware": "1",
        "location": "c5b9744a0db340688588c7f8f1a07e5d",
        "model": "Tom/Floor",
        "name": "Pomp 1 - Kamer Beau",
        "zigbee_mac_address": "000D6F0015476C80",
        "vendor": "Plugwise",
        "available": true,
        "sensors": {
          "temperature": 23.2,
          "setpoint": 16.0,
          "temperature_difference": 2.1,
          "valve_position": 0.0
        }
      },
      "8cb89144d0504e86a0655bd6e9a60b14": {
        "dev_class": "zone_thermostat",
        "firmware": "2016-10-27T02:00:00+02:00",
        "hardware": "255",
        "location": "11d7e32c6ed74582b1700169654c670b",
        "model": "Lisa",
        "name": "Thermostaat - Kamer F",
        "zigbee_mac_address": "000D6F000C86B5A1",
        "vendor": "Plugwise",
        "thermostat": {
          "setpoint": 10.0,
          "lower_bound": 0.0,
          "upper_bound": 99.9,
          "resolution": 0.01
        },
        "available": true,
        "preset_modes": [
          "home",
          "vacation",
          "no_frost",
          "asleep",
          "away"
        ],
        "active_preset": "no_frost",
        "available_schedules": [
          "Thermostat schedule"
        ],
        "selected_schedule": "None",
        "last_used": "Thermostat schedule",
        "control_state": "off",
        "mode": "heat",
        "sensors": {
          "temperature": 20.5,
          "setpoint": 10.0,
          "battery": 54
        }
      },
      "fd2e259f34d54b46824c76e655f571ed": {
        "dev_class": "thermo_sensor",
        "firmware": "2019-03-27T01:00:00+01:00",
        "hardware": "1",
        "location": "11d7e32c6ed74582b1700169654c670b",
        "model": "Tom/Floor",
        "name": "Pomp 4 - Kamer Fleur",
        "zigbee_mac_address": "000D6F0015477039",
        "vendor": "Plugwise",
        "available": true,
        "sensors": {
          "temperature": 23.6,
          "setpoint": 10.0,
          "temperature_difference": -0.1,
          "valve_position": 0.0
        }
      },
      "e1fed7b9d1ee45ab931d7954be86a6a1": {
        "dev_class": "zone_thermostat",
        "firmware": "2016-10-27T02:00:00+02:00",
        "hardware": "255",
        "location": "16666ce849744cb0ab6f900e07dffa9b",
        "model": "Lisa",
        "name": "Thermostaat - Zolder",
        "zigbee_mac_address": "000D6F000C85FD36",
        "vendor": "Plugwise",
        "thermostat": {
          "setpoint": 17.5,
          "lower_bound": 0.0,
          "upper_bound": 99.9,
          "resolution": 0.01
        },
        "available": true,
        "preset_modes": [
          "home",
          "vacation",
          "no_frost",
          "asleep",
          "away"
        ],
        "active_preset": "asleep",
        "available_schedules": [
          "Thermostat schedule"
        ],
        "selected_schedule": "None",
        "last_used": "Thermostat schedule",
        "control_state": "off",
        "mode": "heat",
        "sensors": {
          "temperature": 18.9,
          "setpoint": 17.5,
          "battery": 55
        }
      },
      "12838e10f64d4b4799ff6fb2f5806ef0": {
        "dev_class": "thermo_sensor",
        "firmware": "2020-11-04T01:00:00+01:00",
        "hardware": "1",
        "location": "d721a78bfad04804bab931e81c51ca93",
        "model": "Tom/Floor",
        "name": "Pomp 3 - Master Bedroom",
        "zigbee_mac_address": "000D6F0015980E51",
        "vendor": "Plugwise",
        "available": true,
        "sensors": {
          "temperature": 23.3,
          "setpoint": 16.0,
          "temperature_difference": 2.1,
          "valve_position": 0.0
        }
      },
      "f8031fa0dd8a4ee497d614d8fe1c0c8e": {
        "dev_class": "gateway",
        "firmware": "3.7.0",
        "hardware": "AME Smile 2.0 board",
        "location": "5b11bdc94fdd4be3808801da4c39277b",
        "mac_address": "D40FB201CD50",
        "model": "Gateway",
        "name": "Adam",
        "zigbee_mac_address": "000D6F000D80C361",
        "vendor": "Plugwise",
        "regulation_modes": [
          "heating",
          "off",
          "bleeding_cold",
          "bleeding_hot"
        ],
        "binary_sensors": {
          "plugwise_notification": false
        },
        "sensors": {
          "outdoor_temperature": 10.1
        }
      },
      "fcb48c3d24e74849b259420ffaf017b3": {
        "dev_class": "thermo_sensor",
        "firmware": "2020-11-04T01:00:00+01:00",
        "hardware": "1",
        "location": "d721a78bfad04804bab931e81c51ca93",
        "model": "Tom/Floor",
        "name": "Pomp 2 - Master Bedroom",
        "zigbee_mac_address": "000D6F0015982AF0",
        "vendor": "Plugwise",
        "available": true,
        "sensors": {
          "temperature": 23.3,
          "setpoint": 16.0,
          "temperature_difference": 2.4,
          "valve_position": 0.0
        }
      },
      "72ab8f2b823e49598bcc30551abf0a46": {
        "dev_class": "heater_central",
        "location": "5b11bdc94fdd4be3808801da4c39277b",
        "model": "Generic heater",
        "name": "OpenTherm",
        "vendor": "Intergas Verwarming B.V.",
        "maximum_boiler_temperature": {
          "setpoint": 50.0,
          "lower_bound": 25.0,
          "upper_bound": 95.0,
          "resolution": 0.01
        },
        "domestic_hot_water_setpoint": {
          "setpoint": 60.0,
          "lower_bound": 40.0,
          "upper_bound": 60.0,
          "resolution": 0.01
        },
        "available": true,
        "binary_sensors": {
          "dhw_state": false,
          "heating_state": false,
          "flame_state": false
        },
        "sensors": {
          "water_temperature": 55.2,
          "dhw_temperature": 44.2,
          "intended_boiler_temperature": 0.0,
          "modulation_level": 0.0,
          "water_pressure": 1.84
        },
        "switches": {
          "dhw_cm_switch": true,
          "cooling_ena_switch": false
        }
      },
      "622ef7ff5e9a495694a220ac4957ca26": {
        "dev_class": "zone_thermostat",
        "firmware": "2016-10-27T02:00:00+02:00",
        "hardware": "255",
        "location": "c5b9744a0db340688588c7f8f1a07e5d",
        "model": "Lisa",
        "name": "Thermostaat - Kamer B",
        "zigbee_mac_address": "000D6F001609798A",
        "vendor": "Plugwise",
        "thermostat": {
          "setpoint": 16.0,
          "lower_bound": 0.0,
          "upper_bound": 99.9,
          "resolution": 0.01
        },
        "available": true,
        "preset_modes": [
          "home",
          "vacation",
          "no_frost",
          "asleep",
          "away"
        ],
        "active_preset": "asleep",
        "available_schedules": [
          "Thermostat schedule"
        ],
        "selected_schedule": "None",
        "last_used": "Thermostat schedule",
        "control_state": "off",
        "mode": "heat",
        "sensors": {
          "temperature": 20.1,
          "setpoint": 16.0,
          "battery": 56
        }
      },
      "320734ee456640f087ce7af04a5bfabe": {
        "dev_class": "zz_misc",
        "location": "16666ce849744cb0ab6f900e07dffa9b",
        "model": "lumi.plug.maeu01",
        "name": "Pomp 5 - Zolder",
        "zigbee_mac_address": "54EF4410000FF203",
        "vendor": "LUMI",
        "available": true,
        "sensors": {
          "electricity_consumed_interval": 0.0
        },
        "switches": {
          "relay": false,
          "lock": false
        }
      },
      "bea5074ef7ff4ef199ae37e18c5848e6": {
        "dev_class": "zone_thermostat",
        "firmware": "2016-10-10T02:00:00+02:00",
        "hardware": "255",
        "location": "d721a78bfad04804bab931e81c51ca93",
        "model": "Lisa",
        "name": "Thermostaat - Master Bedroom",
        "zigbee_mac_address": "000D6F000C86BAEC",
        "vendor": "Plugwise",
        "thermostat": {
          "setpoint": 16.0,
          "lower_bound": 0.0,
          "upper_bound": 99.9,
          "resolution": 0.01
        },
        "available": true,
        "preset_modes": [
          "home",
          "vacation",
          "no_frost",
          "asleep",
          "away"
        ],
        "active_preset": "asleep",
        "available_schedules": [
          "Thermostat schedule"
        ],
        "selected_schedule": "None",
        "last_used": "Thermostat schedule",
        "control_state": "off",
        "mode": "heat",
        "sensors": {
          "temperature": 18.9,
          "setpoint": 16.0,
          "battery": 63
        }
      }
    }
  }
}
```

**Comment by bouwew** [comment]
Thanks!

I see that the `regulation_mode` key is indeed missing.
Very strange.

Can you please have a look at the data on your Adam, look here:` http://<adam_ip>/core/appliances` (you need to log in with username = smile and password = 8-letter device ID)
You should see xml-data, look for `regulation_mode`, you should see something similar like this:
![image](https://user-images.githubusercontent.com/11290930/200244906-3cd39107-a6ed-4729-83d7-aaa9e9fefb81.png)
And a little more down there should be this:
![image](https://user-images.githubusercontent.com/11290930/200245046-49869b22-3bac-4c4a-9bff-3643d4914be9.png)

I wonder how this data looks on your Adam? I'm suspecting what's shown in the first picture is missing or not complete.

**Comment by Xirt** [comment]
The first section looks like this:

```
<point_log id="3f91d63c2232482ca3e600f060e8dc0f">
<type>regulation_mode</type>
<unit/>
<updated_date/>
<last_consecutive_log_date/>
<interval/>
<regulation_mode_control id="39aabbc1ab7f47d3a51b65c2632d9ca6"/>
</point_log>
```

The second section looks as follows:
```
<regulation_mode_control_functionality id="716d5970c9f4421e9eae8ef913d3ade0">
<regulation_mode_control id="39aabbc1ab7f47d3a51b65c2632d9ca6"/>
<updated_date/>
<type>regulation_mode</type>
<mode>heating</mode>
<allowed_modes>
<mode>off</mode>
<mode>bleeding_hot</mode>
<mode>bleeding_cold</mode>
<mode>heating</mode>
</allowed_modes>
</regulation_mode_control_functionality>
```

**Comment by bouwew** [comment]
Ok, yes, that's the problem. 
I had anticipated this and already programmed a fix. Can you try this fix with plugwise-beta?
If yes, I'll make a test-release during the weekend.

Another option would be for you to contact Plugwise Helpdesk, ask them to reset the database on your Adam, because the existing database is corrupted. You can link this issue in the email for illustration.

**Comment by Xirt** [comment]
I configured the plugwise-beta version in a temporary container, but unfortunately no joy (unless I am doing something wrong). See below the log entry:

```
2022-11-11 01:09:51.473 ERROR (MainThread) [homeassistant.components.select] Error adding entities for domain select with platform plugwise
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 428, in async_add_entities
await asyncio.gather(*tasks)
File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 692, in _async_add_entity
await entity.add_to_platform_finish()
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 788, in add_to_platform_finish
await self.async_added_to_hass()
File "/config/custom_components/plugwise/entity.py", line 81, in async_added_to_hass
self._handle_coordinator_update()
File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 348, in _handle_coordinator_update
self.async_write_ha_state()
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 545, in async_write_ha_state
self._async_write_ha_state()
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 583, in _async_write_ha_state
state = self._stringify_state(available)
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 551, in _stringify_state
if (state := self.state) is None:
File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 97, in state
if self.current_option is None or self.current_option not in self.options:
File "/config/custom_components/plugwise/select.py", line 120, in current_option
return self.device[self.entity_description.current_option_key]
KeyError: 'regulation_mode'
2022-11-11 01:09:51.481 ERROR (MainThread) [homeassistant.components.select] Error while setting up plugwise platform for select
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 289, in _async_setup_platform
await asyncio.gather(*pending)
File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 428, in async_add_entities
await asyncio.gather(*tasks)
File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 692, in _async_add_entity
await entity.add_to_platform_finish()
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 788, in add_to_platform_finish
await self.async_added_to_hass()
File "/config/custom_components/plugwise/entity.py", line 81, in async_added_to_hass
self._handle_coordinator_update()
File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 348, in _handle_coordinator_update
self.async_write_ha_state()
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 545, in async_write_ha_state
self._async_write_ha_state()
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 583, in _async_write_ha_state
state = self._stringify_state(available)
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 551, in _stringify_state
if (state := self.state) is None:
File "/usr/src/homeassistant/homeassistant/components/select/__init__.py", line 97, in state
if self.current_option is None or self.current_option not in self.options:
File "/config/custom_components/plugwise/select.py", line 120, in current_option
return self.device[self.entity_description.current_option_key]
KeyError: 'regulation_mode'
```

**Comment by bouwew** [comment]
That correct, you are testing too soon :)
I wrote, I will make a test-release during the weekend, the fix hasn't been implemented in any release yet.

**Comment by Xirt** [comment]
No issue. I now have a test box / container ready for the job ;-). Thanks for the update and please let me know once you have the release ready.

**Comment by bouwew** [comment]
Please test with plugwise-beta version 0.31.3a3

**Comment by Xirt** [comment]
Just tested fine: all information I use in dashboards appear to show fine, some quick configurations (tests) of temperatures work fine and no errors in the log :).

**Comment by bouwew** [comment]
Nice! Thanks for testing!

**Comment by bouwew** [comment]
Should be fixed in HA Core 2022.12.0.
